llvm/tools/llvm-foreach/llvm-foreach.cpp

```diff
@@ -0,0 +1,189 @@
+//===- llvm-foreach.cpp - Command lines execution ---------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This source is utility to execute command lines. The specified command will
+// be invoked as many times as necessary to use up the list of input items.
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/LineIterator.h"
+#include "llvm/Support/MemoryBuffer.h"
+#include "llvm/Support/Path.h"
+#include "llvm/Support/Program.h"
+#include "llvm/Support/SystemUtils.h"
+
+#include <vector>
+
+using namespace llvm;
+
+static cl::list<std::string> InputFileLists{"in-file-list", cl::OneOrMore,
+                                            cl::desc("<input list of files>"),
+                                            cl::value_desc("filename")};
+
+static cl::list<std::string> InputCommandArgs{
+    cl::Positional, cl::OneOrMore, cl::desc("<command>"),
+    cl::value_desc("command")};
+
+static cl::list<std::string> Replaces{
+    "in-replace", cl::OneOrMore,
+    cl::desc("Specify input path in input command, this will be replaced with "
+             "names read from corresponding input list of files;"),
+    cl::value_desc("R")};
+
+static cl::opt<std::string> OutReplace{
+    "out-replace",
+    cl::desc("Specify output path in input command, this will be replaced with "
+             "name of temporary file created for writing command's outputs."),
+    cl::init(""), cl::value_desc("R")};
+
+static cl::opt<std::string> OutFilesExt{
```

Comment:
This is usually done by a command itself, e.g. `echo 1.txt 2.txt 3.txt | xargs -I{} cp {} {}.copy`

Reply:
> To simplify the tool, I tried to avoid searching a substring in strings, this approach will force me to do it. Are you sure?

I'm not sure I understand the reason why you tried to avoid searching for a substring.
Do you want to only support a case when a pattern (e.g. "{}" in the example above) is used as a complete argument, and avoid substitution when {} is a substring of an argument?

Reply:
It was enough for my cases :)
But since I needed support arguments with `=` symbol, maybe it's time to search substring in arguments strings.

Reply:
Although I'm not sure how are you going to make the "output files list" option. How this foreach tool will understand that some argument is an input, and some is an output? What if a command does not have any "output" files, or it makes transformation in place? What if a command has *two* or more output files for each input files?

Xargs (as well as `find -exec`) does not have a notion of output (or input) files. All they handle is a command, and string substitution. In case of `llvm-foreach`, having an output files list makes things complicated or a lot less flexible, if you don't support these cases.

Reply:
> Although I'm not sure how are you going to make the "output files list" option. How this foreach tool understands that some argument is an input, and another is an output? What if a command does not have any "output" files, or it makes transformation in place? What if a command has _two_ or more output files for each input files?

I ask user to provide different substitutions for input and output files through `-in-replace` and `-out-replace` options.
Two and more output files from a command is a rare situation, so `llvm-foreach` doesn't support this case yet, but I think it can be extended like I extended it for multiple inputs for command.

> Xargs (as well as `find -exec`) does not have a notion of output (or input) files. All they handle is a command, and string substitution. In case of `llvm-foreach`, having an output files list makes things complicated or a lot less flexible, if you don't support these cases.

All I'm doing in this tool - I'm fighting with clang driver limitations :( Driver can only handle file lists, so I taught `llvm-foreach` emit file lists. 
As I said, I'll add searching of substitutions in arg strings.

Reply:
> I'm fighting with clang driver limitations :( Driver can only handle file lists, so I taught `llvm-foreach` emit file lists.

Ok, ok :)
The tool seems to be a acceptable workaround, until someone can figure out how to improve the driver.

Reply:
Added substitutions, removed unnecessary option.

Reply:
@asavonic , it looks like I can't just remove this option for now. If extension added by the command, llvm-foreach misses extension and writes filenames without extensions to output file list. Not sure that I want search for extension by llvm-foreach in resulting arguments. + Adding extension by command makes already complicated driver's code even more complicated.
So, If you don't mind, I'll leave the `-out-ext` option.